File: LinkedListCycle.java
public class LinkedListCycle {
    //141. Linked List Cycle
    //https://leetcode.com/problems/linked-list-cycle/description/
    public boolean hasCycle(ListNode head) {
        //Use Fast and Slow Pointer.
        //Initalize Fast and Slow pointer to start at the head.
        //Fast will by two nodes. Slow will by one node.
        //If both fast and slow pointer meet, then we have a cycle. Return true. 
        //Otherwise, return false if either fast or fast.next is null.

        if(head == null){
            return false;
        }

        ListNode fast = head;
        ListNode slow = head;

        while(fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;
            if(fast == slow){
                return true;
            }
        }

        return false;
    }
}

File: LinkedListCycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
#Use Fast and Slow Pointer
#Init fast and slow to head node.
#Fast moves by two nodes
#Slow moves by one node.
#If both the fast and slow pointer meet at the same node, return true.
#Otherwise return false if fast or fast.next is None.
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False

        fast = head
        slow = head

        while fast is not None and fast.next is not None:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                return True
        
        return False